lxd/instance_state.go

```diff
@@ -65,145 +64,113 @@ func containerStatePut(d *Daemon, r *http.Request) response.Response {
-				c.SetOperation(op)
-				if c.IsFrozen() {
-					err := c.Unfreeze()
-					if err != nil {
-						return err
-					}
-				}
-
-				err = c.Shutdown(time.Duration(raw.Timeout) * time.Second)
-				if err != nil {
-					return err
-				}
-
-				return nil
-			}
+			return inst.Shutdown(time.Duration(req.Timeout) * time.Second)
 		}
 	case shared.Restart:
-		opType = db.OperationInstanceRestart
-		do = func(op *operations.Operation) error {
-			c.SetOperation(op)
-			ephemeral := c.IsEphemeral()
-
-			if ephemeral {
-				// Unset ephemeral flag
-				args := db.InstanceArgs{
-					Architecture: c.Architecture(),
-					Config:       c.LocalConfig(),
-					Description:  c.Description(),
-					Devices:      c.LocalDevices(),
-					Ephemeral:    false,
-					Profiles:     c.Profiles(),
-					Project:      c.Project(),
-					Type:         c.Type(),
-					Snapshot:     c.IsSnapshot(),
-				}
-
-				err := c.Update(args, false)
-				if err != nil {
-					return err
-				}
-
-				// On function return, set the flag back on
-				defer func() {
-					args.Ephemeral = ephemeral
-					c.Update(args, false)
-				}()
+		ephemeral := inst.IsEphemeral()
+
+		if ephemeral {
+			// Unset ephemeral flag
+			args := db.InstanceArgs{
+				Architecture: inst.Architecture(),
+				Config:       inst.LocalConfig(),
+				Description:  inst.Description(),
+				Devices:      inst.LocalDevices(),
+				Ephemeral:    false,
+				Profiles:     inst.Profiles(),
+				Project:      inst.Project(),
+				Type:         inst.Type(),
+				Snapshot:     inst.IsSnapshot(),
 			}
 
-			timeout := raw.Timeout
-			if raw.Force {
-				timeout = 0
+			err := inst.Update(args, false)
+			if err != nil {
+				return err
 			}
-			return c.Restart(time.Duration(timeout))
-		}
-	case shared.Freeze:
-		if !d.os.CGInfo.Supports(cgroup.Freezer, nil) {
-			return response.BadRequest(fmt.Errorf("This system doesn't support freezing instances"))
-		}
 
-		opType = db.OperationInstanceFreeze
-		do = func(op *operations.Operation) error {
-			c.SetOperation(op)
-			return c.Freeze()
-		}
-	case shared.Unfreeze:
-		if !d.os.CGInfo.Supports(cgroup.Freezer, nil) {
-			return response.BadRequest(fmt.Errorf("This system doesn't support unfreezing instances"))
+			// On function return, set the flag back on
+			defer func() {
+				args.Ephemeral = ephemeral
+				inst.Update(args, false)
+			}()
 		}
 
-		opType = db.OperationInstanceUnfreeze
-		do = func(op *operations.Operation) error {
-			c.SetOperation(op)
-			return c.Unfreeze()
+		timeout := req.Timeout
+		if req.Force {
+			timeout = 0
 		}
-	default:
-		return response.BadRequest(fmt.Errorf("Unknown action %s", raw.Action))
-	}
-
-	resources := map[string][]string{}
-	resources["instances"] = []string{name}
 
-	op, err := operations.OperationCreate(d.State(), project, operations.OperationClassTask, opType, resources, nil, do, nil, nil)
-	if err != nil {
-		return response.InternalError(err)
+		return inst.Restart(time.Duration(timeout))
```

Comment:
@stgraber can the detection of the Ephemeral property and calls to Update() be moved into `Restart()` functions?

Reply:
Yeah, I was thinking that too. We didn't use to have a Restart() function but now that we do, this can be moved over.